### ISSUE TITLE
Sgen aborting because of theorically unreachable code

### ISSUE THREAD

**Issue by tmijieux** [mono/mono#21369]
potentially similar issue ? https://github.com/mono/mono/issues/14732

TL;DR: I think I have found a workaround for now by using the "new" sgen bridge implementation.
But i am curious if someone here with good knowledge of the sgen garbage / bridge can give me insight 
of what is going on with this issue: 
- https://github.com/xamarin/xamarin-android/issues/6546

I already gave a lot of information in the xamarin-android issue so do not hesitate to read that first.

## Steps to Reproduce

could not yet reproduce with a minimal reproduction project (EDIT: see [response below](https://github.com/mono/mono/issues/21369#issuecomment-1001189366), i now found out a way to trigger the bug )

### Current Behavior

crash because of sgen gc aborting

### Expected Behavior

no crash

## On which platforms did you notice this

[ ] macOS
[ ] Linux
[ ] Windows
[X] Android with tarjan bridge and xamarin for android (not .NET 6, regular "old" xamarin application)

**Version Used**:

mono-6.12.0.145 on May 27  2021 c633fe9
(I know this is not the latest, in the eventuality that there is a problem that have been identified since, and i did not find the issue, then, sorry and please point me to relevant PR/issues)

```
#0  0x00007c72ab4f02a8 in syscall () from C:\Users\Thomas\AppData\Local\Temp\x64\Debug\.gdb\libc.so
#1  0x00007c72ab4f3213 in abort () from C:\Users\Thomas\AppData\Local\Temp\x64\Debug\.gdb\libc.so
#2  0x00007c6fbf23174a in ?? ()
#3  0x00007c6fbf1d9cf0 in ?? ()
#4  0x00007c6fbfb0f154 in eglib_log_adapter (log_domain=0x0, log_level=G_LOG_LEVEL_ERROR, message=0x7c70c92a0e90 "* Assertion: should not be reached at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-scan-object.h:91\n", user_data=0x0) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/utils/mono-logger.c:405
#5  0x00007c6fbfb3c72a in monoeg_g_logstr (log_domain=0x0, log_level=G_LOG_LEVEL_ERROR, msg=0x7c70c92a0e90 "* Assertion: should not be reached at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-scan-object.h:91\n") at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/eglib/goutput.c:151
#6  0x00007c6fbfb3c00e in monoeg_g_logv_nofree (log_domain=0x0, log_level=G_LOG_LEVEL_ERROR, format=0x7c6fbfb96dca "* Assertion: should not be reached at %s:%d\n", args=0x7c6fbf1d9020) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/eglib/goutput.c:166
#7  0x00007c6fbfb3c2f4 in monoeg_assertion_message (format=0x7c6fbfb96dca "* Assertion: should not be reached at %s:%d\n") at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/eglib/goutput.c:201
#8  0x00007c6fbfb3c394 in mono_assertion_message_unreachable (file=0x7c6fbfb8bb6a "/Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-scan-object.h", line=91) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/eglib/goutput.c:228
#9  0x00007c6fbfaa0c77 in major_scan_object_concurrent_with_evacuation (full_object=0x7c6fb7a157e0, desc=35027181184716800, queue=0x7c72aac6b010) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-scan-object.h:91
#10 0x00007c6fbfac3dd1 in scan_card_table_for_block (block=0x7c6fb7a14000, scan_type=CARDTABLE_SCAN_MOD_UNION_PRECLEAN, ctx=...) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-marksweep.c:2619
#11 0x00007c6fbfa91e24 in major_scan_card_table (scan_type=CARDTABLE_SCAN_MOD_UNION_PRECLEAN, ctx=..., job_index=0, job_split_count=1, block_count=4973) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-marksweep.c:2711
#12 0x00007c6fbfa877d2 in job_major_mod_union_preclean (worker_data_untyped=0x7c72aac6b008, job=0x7c7001dc3208) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-gc.c:1554
#13 0x00007c6fbfb00947 in thread_func (data=0x0) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-thread-pool.c:207
#14 0x00007c72ab55dd2b in __pthread_start(void*) () from C:\Users\Thomas\AppData\Local\Temp\x64\Debug\.gdb\libc.so
#15 0x00007c72ab4f50c8 in __start_thread () from C:\Users\Thomas\AppData\Local\Temp\x64\Debug\.gdb\libc.so
```

<details>
<summary>variant with "large object" (not meaning "in LOS" , rather bigger than card in cardtable)</summary>
<p>

```
#0  0x00007018985dd2a8 in syscall () from C:\Users\tmijieux\AppData\Local\Temp\x64\Debug\.gdb\libc.so
#1  0x00007018985e0213 in abort () from C:\Users\tmijieux\AppData\Local\Temp\x64\Debug\.gdb\libc.so
#2  0x00007015ac12f68a in ?? ()
#3  0x00007015abbfbcf0 in ?? ()
#4  0x00007015aca11154 in eglib_log_adapter (log_domain=0x0, log_level=G_LOG_LEVEL_ERROR, message=0x7016b6004970 "* Assertion: should not be reached at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-scan-object.h:91\n", user_data=0x0) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/utils/mono-logger.c:405
#5  0x00007015aca3e72a in monoeg_g_logstr (log_domain=0x0, log_level=G_LOG_LEVEL_ERROR, msg=0x7016b6004970 "* Assertion: should not be reached at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-scan-object.h:91\n") at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/eglib/goutput.c:151
#6  0x00007015aca3e00e in monoeg_g_logv_nofree (log_domain=0x0, log_level=G_LOG_LEVEL_ERROR, format=0x7015aca98dca "* Assertion: should not be reached at %s:%d\n", args=0x7015abbfafc0) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/eglib/goutput.c:166
#7  0x00007015aca3e2f4 in monoeg_assertion_message (format=0x7015aca98dca "* Assertion: should not be reached at %s:%d\n") at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/eglib/goutput.c:201
#8  0x00007015aca3e394 in mono_assertion_message_unreachable (file=0x7015aca8db6a "/Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-scan-object.h", line=91) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/eglib/goutput.c:228
#9  0x00007015ac9a2c77 in major_scan_object_concurrent_with_evacuation (full_object=0x7015a4797c40, desc=31550443109896192, queue=0x701897c56010) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-scan-object.h:91
#10 0x00007015ac96f0f5 in sgen_cardtable_scan_object (obj=0x7015a4797c40, block_obj_size=840, cards=0x7015abbfba4e "", ctx=...) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-cardtable.c:590
#11 0x00007015ac9c5e13 in scan_card_table_for_block (block=0x7015a4794000, scan_type=CARDTABLE_SCAN_MOD_UNION_PRECLEAN, ctx=...) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-marksweep.c:2622
#12 0x00007015ac993e24 in major_scan_card_table (scan_type=CARDTABLE_SCAN_MOD_UNION_PRECLEAN, ctx=..., job_index=0, job_split_count=1, block_count=4860) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-marksweep.c:2711
#13 0x00007015ac9897d2 in job_major_mod_union_preclean (worker_data_untyped=0x701897c56008, job=0x7015ad9ae008) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-gc.c:1554
#14 0x00007015aca02947 in thread_func (data=0x0) at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/debug/mono/sgen/sgen-thread-pool.c:207
#15 0x000070189864ad2b in __pthread_start(void*) () from C:\Users\tmijieux\AppData\Local\Temp\x64\Debug\.gdb\libc.so
#16 0x00007018985e20c8 in __start_thread () from C:\Users\tmijieux\AppData\Local\Temp\x64\Debug\.gdb\libc.so

```

</p>
</details>

Like i said in the xamarin-android issue, my first guess was maybe memory corruption because of any bug or other native library. 
But then i read little bits of code and documentation about sgen. 
By inspecting the values in the debugger in noticed that the vtable pointer is ending with 7 so i am wondering if there could be a corner case race condition because of tarjan where object would get re-tagged by bridge and then read on scan function on a place where it does not expect it to be tagged ?

**Comment by tmijieux** [comment]
after resetting the last 3 lower bits of vtable pointer to zero (in debugger). I can see that it is probably  NOT some kind of memory corruption because I can see regular objects (of various type among the different reproduction)

in the COPY_OR_MARK function there is these few lines
https://github.com/mono/mono/blob/1734a7d60bd156cf78a0d9745557ed73c4f95055/mono/sgen/sgen-marksweep-drain-gray-stack.h#L156-L160

that seems to indicate there could be the bridge running concurrently to the copy_or_mark func, but in the code that get the invalid descriptor that trigger the bug here
https://github.com/mono/mono/blob/1734a7d60bd156cf78a0d9745557ed73c4f95055/mono/sgen/sgen-marksweep.c#L2661-L2665 
there is the regular call to `sgen_obj_get_descriptor` that does not do the same, (there is a assert check but level is 9 so i guess the it is skipped in regular production build with MAX_ASSERT_LEVEL to 6)

is that normal ?

**Comment by tmijieux** [comment]
what i am observing is clearly the tagging from the tarjan bridge because the "regular" tagging cannot have the three tags : forwarded , pinned and cemented at the same time.

This is not used in the other bridge processor implementations so that comforts me that changing the bridge implementation is a valid workaround for preventing this bug.

https://github.com/mono/mono/blob/73a3a8adb2b39b0598dab3fec135b46f2bf19fc5/mono/metadata/sgen-tarjan-bridge.c#L342-L358

thus it seems to indicate that the `bridge_processing_stw_step()` (bridge processing step during world is stopped) is running in parallel with the `job_major_mod_union_preclean()` which a worker job enqueued in the finish callback of the workers `workers_finish_callback()` (from my undestanding this job is supposed to be called one time when all other worker jobs have completed)

https://github.com/mono/mono/blob/73a3a8adb2b39b0598dab3fec135b46f2bf19fc5/mono/sgen/sgen-gc.c#L1622-L1634
https://github.com/mono/mono/blob/73a3a8adb2b39b0598dab3fec135b46f2bf19fc5/mono/sgen/sgen-gc.c#L2179-L2184

the `workers_finish_callback` is setup when starting the workers in `major_copy_or_mark_from_roots()` in `major_start_collection()`
the `bridge_processing_stw_step()` is called from finish_gray_stack that is called from `major_finish_collection()`

the workers finish callback, major_start_collection and major_finish_collection could all be called from multiple points in code, so i am not yet sure what path was taken here, in some paths there are stuff like `sgen_workers_stop_all_workers()` or `sgen_workers_start_all_workers()` with a null finish callback so I did not understand yet what is happening exactly, but i guess there must be a path/race condition where this can happen. I will try to narrow it down (maybe the binary protocol stuff can help me trace what is happening, I did not try it yet).

**Comment by tmijieux** [comment]
[binprot.log](https://github.com/mono/mono/files/7775739/binprot.log)

the binary protocol log at the time the crash happens

**Comment by tmijieux** [comment]
i added a little bit of code in the sgen and tarjan bridge code to add condition for them to wait for each other here
(and a dummy bridge implementation):

https://github.com/tmijieux/mono/commit/e9cd7fbe24502e4c57b3d6aac29cb100dc6a9b0d

<details>
<summary> simple program.cs to reproduce bug</summary>

```cs
using System;

namespace MyApp
{
    public class OtherStuff
    {
        public double[] lel;
        public OtherStuff()
        {
            lel = new double[10];
            for (int i = 0; i < 10; ++i){
                lel[i] = i;
            }
        }

        // need finalizer to be considered bridge
        ~OtherStuff()
        {
            lel[0] = lel[3]+lel[4];
        }
    }

    public class MyStuff
    {
        public OtherStuff oa;
        public OtherStuff ob;
        public MyStuff prev;


        public int k;

        public MyStuff()
        {
            oa = new OtherStuff();
            ob = new OtherStuff();
            k = (int)(ob.lel[3] + oa.lel[7]);
        }

        // need finalizer to be considered bridge
        ~MyStuff()
        {
            oa = ob;
            prev.oa = null;
        }
    }

    public class Program
    {
        public static void Main(string[] args)
        {

            Console.WriteLine("Hello World!");
            MyStuff prev = null;
            while (true)
            {
                int  k = 0;
                for (int i = 0; i < 5000; ++i)
                {
                    var n = new MyStuff();
                    n.prev = prev;
                    prev = n;
                    k += n.k;
                }
                prev = null;
                Console.WriteLine($"Tick ! k={k}");
            }

        }
    }
}

```

</details>


If you run this patch on this very simple csharp code, then the bug will reproduce systematically ! (on regular linux box which is more friendly to debug than android simulator/device):


<details>
<summary> crash log</summary>

```
$ MONO_GC_PARAMS="bridge-implementation=tarjan" MONO_LOG_LEVEL="error" MONO_LOG_MASK="gc" /opt/mono/bin/mono-sgen ~/lel/Program.exe
my debug init!
Hello World!
Tick ! k=50000
Tick ! k=50000
bridge stw enter !
bridge left !
Tick ! k=50000
Tick ! k=50000
Tick ! k=50000
bridge stw enter !
bridge left !
Tick ! k=50000
Tick ! k=50000
bridge stw enter !
bridge left !
Tick ! k=50000
Tick ! k=50000
Tick ! k=50000
bridge stw enter !
bridge left !
Tick ! k=50000
Tick ! k=50000
bridge stw enter !
bridge left !
signaling bridge_cond (can_leave=FALSE)
waiting for job_cond!
Tick ! k=50000
Tick ! k=50000
Tick ! k=50000
bridge stw enter !
set can start job !
signaling job cond !
waiting for bridge cond !
starting job_major_mod_union_preclean
* Assertion: should not be reached at sgen-scan-object.h:91


=================================================================
        Native Crash Reporting
=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================

=================================================================
        Native stacktrace:
=================================================================
        0x55a80f845a11 - /opt/mono/bin/mono-sgen :
        0x55a80f845da9 - /opt/mono/bin/mono-sgen :
        0x55a80f7f37d1 - /opt/mono/bin/mono-sgen :
        0x55a80f844fa4 - /opt/mono/bin/mono-sgen :
        0x7f91df107870 - /usr/lib/libpthread.so.0 :
        0x7f91def47d22 - /usr/lib/libc.so.6 : gsignal
        0x7f91def31862 - /usr/lib/libc.so.6 : abort
        0x55a80f7b60d0 - /opt/mono/bin/mono-sgen :
        0x55a80faaa9c5 - /opt/mono/bin/mono-sgen :
        0x55a80fac9901 - /opt/mono/bin/mono-sgen : monoeg_assertion_message
        0x55a80fac9987 - /opt/mono/bin/mono-sgen :
        0x55a80fa82e96 - /opt/mono/bin/mono-sgen :
        0x55a80fa7d069 - /opt/mono/bin/mono-sgen :
        0x55a80fa7d2b0 - /opt/mono/bin/mono-sgen :
        0x55a80fa6d1b6 - /opt/mono/bin/mono-sgen :
        0x55a80faa16f6 - /opt/mono/bin/mono-sgen :
        0x7f91df0fd259 - /usr/lib/libpthread.so.0 :
        0x7f91df0095e3 - /usr/lib/libc.so.6 : clone

=================================================================
        Telemetry Dumper:
=================================================================
Thread 0x7f91de3ff640 starting summarize_execute
Pkilling 0x7f91dc792640 from 0x7f91de3ff640
Pkilling 0x7f91def08740 from 0x7f91de3ff640
Entering thread summarizer pause from 0x7f91deaef640
Finished thread summarizer pause from 0x7f91deaef640.
domain required for stack walkFailed to create breadcrumb file (null)/crash_hash_0x0

Waiting for dumping threads to resume

=================================================================
        External Debugger Dump:
=================================================================
[New LWP 164146]
[New LWP 164147]
[New LWP 164148]
warning: File "/opt/mono/bin/mono-sgen-gdb.py" auto-loading has been declined by your `auto-load safe-path' set to "$debugdir:$datadir/auto-load".
To enable execution of this file add
        add-auto-load-safe-path /opt/mono/bin/mono-sgen-gdb.py
line to your configuration file "/home/tmijieux/.config/gdb/gdbinit".
To completely disable this security protection add
        set auto-load safe-path /
line to your configuration file "/home/tmijieux/.config/gdb/gdbinit".
For more information about this security protection see the
"Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
        info "(gdb)Auto-loading safe path"
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
0x00007f91df1098ca in __futex_abstimed_wait_common64 () from /usr/lib/libpthread.so.0
  Id   Target Id                                            Frame
* 1    Thread 0x7f91def08740 (LWP 164145) "mono-sgen"       0x00007f91df1098ca in __futex_abstimed_wait_common64 () from /usr/lib/libpthread.so.0
  2    Thread 0x7f91de3ff640 (LWP 164146) "SGen worker"     0x00007f91defd6a2f in wait4 () from /usr/lib/libc.so.6
  3    Thread 0x7f91deaef640 (LWP 164147) "Crash Report Le" 0x00007f91df1098ca in __futex_abstimed_wait_common64 () from /usr/lib/libpthread.so.0
  4    Thread 0x7f91dc792640 (LWP 164148) "Finalizer"       0x00007f91df106600 in __lll_lock_wait () from /usr/lib/libpthread.so.0

Thread 4 (Thread 0x7f91dc792640 (LWP 164148) "Finalizer"):
#0  0x00007f91df106600 in __lll_lock_wait () from /usr/lib/libpthread.so.0
#1  0x00007f91df0ff503 in pthread_mutex_lock () from /usr/lib/libpthread.so.0
#2  0x000055a80fa731a9 in mono_os_mutex_lock (mutex=0x55a80fc374a0 <sgen_gc_mutex>) at ../../mono/utils/mono-os-mutex.h:105
#3  mono_coop_mutex_lock (mutex=0x55a80fc374a0 <sgen_gc_mutex>) at ../../mono/utils/mono-coop-mutex.h:57
#4  sgen_gc_lock () at sgen-gc.c:3948
#5  0x000055a80fa46028 in mono_gc_skip_thread_changing (skip=0) at sgen-mono.c:2247
#6  0x000055a80faba0b3 in mono_thread_info_set_flags (flags=MONO_THREAD_INFO_FLAGS_NONE) at mono-threads.c:865
#7  0x000055a80fa3d71e in finalizer_thread (unused=unused@entry=0x0) at gc.c:928
#8  0x000055a80f9f4b6c in start_wrapper_internal (stack_ptr=0x7f91dc793000, start_info=0x0) at threads.c:1238
#9  start_wrapper (data=0x55a810d321b0) at threads.c:1298
#10 0x00007f91df0fd259 in start_thread () from /usr/lib/libpthread.so.0
#11 0x00007f91df0095e3 in clone () from /usr/lib/libc.so.6

Thread 3 (Thread 0x7f91deaef640 (LWP 164147) "Crash Report Le"):
#0  0x00007f91df1098ca in __futex_abstimed_wait_common64 () from /usr/lib/libpthread.so.0
#1  0x00007f91df105ae8 in __new_sem_wait_slow64.constprop.0 () from /usr/lib/libpthread.so.0
#2  0x000055a80f9f61f7 in mono_os_sem_wait (flags=MONO_SEM_FLAGS_ALERTABLE, sem=0x55a80fc2b190 <summarizer_leader_data+16>) at ../../mono/utils/mono-os-semaphore.h:204
#3  summarizer_leader () at threads.c:6663
#4  0x000055a80f9f4b6c in start_wrapper_internal (stack_ptr=0x7f91deaf0000, start_info=0x0) at threads.c:1238
#5  start_wrapper (data=0x55a810d07500) at threads.c:1298
#6  0x00007f91df0fd259 in start_thread () from /usr/lib/libpthread.so.0
#7  0x00007f91df0095e3 in clone () from /usr/lib/libc.so.6

Thread 2 (Thread 0x7f91de3ff640 (LWP 164146) "SGen worker"):
#0  0x00007f91defd6a2f in wait4 () from /usr/lib/libc.so.6
#1  0x000055a80f845c91 in dump_native_stacktrace (signal=signal@entry=0x55a80fae92b5 "SIGABRT", mctx=mctx@entry=0x7f91de3fe040) at mini-posix.c:1075
#2  0x000055a80f845da9 in mono_dump_native_crash_info (signal=signal@entry=0x55a80fae92b5 "SIGABRT", mctx=mctx@entry=0x7f91de3fe040, info=info@entry=0x7f91de3fe330) at mini-posix.c:1119
#3  0x000055a80f7f37d1 in mono_handle_native_crash (signal=0x55a80fae92b5 "SIGABRT", mctx=mctx@entry=0x7f91de3fe040, info=info@entry=0x7f91de3fe330) at mini-exceptions.c:3463
#4  0x000055a80f844fa4 in sigabrt_signal_handler (_dummy=-566239168, _info=0x7f91de3fe330, context=0x7f91de3fe200) at mini-posix.c:247
#5  <signal handler called>
#6  0x00007f91def47d22 in raise () from /usr/lib/libc.so.6
#7  0x00007f91def31862 in abort () from /usr/lib/libc.so.6
#8  0x000055a80f7b60d0 in monoeg_assert_abort () at goutput.c:57
#9  0x000055a80faaa9c5 in mono_log_write_logfile (log_domain=<optimized out>, level=<optimized out>, hdr=<optimized out>, message=0x7f91cc000b60 "* Assertion: should not be reached at sgen-scan-object.h:91\n") at mono-log-common.c:136
#10 0x000055a80fac9901 in monoeg_g_logstr (msg=<optimized out>, log_level=G_LOG_LEVEL_ERROR, log_domain=0x0) at goutput.c:151
#11 monoeg_g_logv_nofree (args=0x7f91de3feb00, format=0x55a80fb958d8 "* Assertion: should not be reached at %s:%d\n", log_level=G_LOG_LEVEL_ERROR, log_domain=0x0) at goutput.c:166
#12 monoeg_g_logv_nofree (args=0x7f91de3feb00, format=0x55a80fb958d8 "* Assertion: should not be reached at %s:%d\n", log_level=G_LOG_LEVEL_ERROR, log_domain=0x0) at goutput.c:155
#13 monoeg_assertion_message (format=format@entry=0x55a80fb958d8 "* Assertion: should not be reached at %s:%d\n") at goutput.c:207
#14 0x000055a80fac9987 in mono_assertion_message_unreachable (file=file@entry=0x55a80fb8b155 "sgen-scan-object.h", line=line@entry=91) at goutput.c:234
#15 0x000055a80fa82e96 in major_scan_object_concurrent_with_evacuation (full_object=full_object@entry=0x7f91d7bc5048, desc=<optimized out>, queue=queue@entry=0x7f91df2b0010) at /home/tmijieux/mono/mono/sgen/sgen-scan-object.h:91
#16 0x000055a80fa7d069 in scan_card_table_for_block (block=0x7f91d7bc4000, scan_type=scan_type@entry=CARDTABLE_SCAN_MOD_UNION_PRECLEAN, ctx=...) at sgen-marksweep.c:2664
#17 0x000055a80fa7d2b0 in major_scan_card_table (scan_type=CARDTABLE_SCAN_MOD_UNION_PRECLEAN, ctx=..., job_index=<optimized out>, job_split_count=<optimized out>, block_count=<optimized out>) at sgen-marksweep.c:2745
#18 0x000055a80fa6d1b6 in job_major_mod_union_preclean (worker_data_untyped=0x7f91df2b0008, job=0x7f91d7a42008) at sgen-gc.c:1604
#19 0x000055a80faa16f6 in thread_func (data=<optimized out>) at sgen-thread-pool.c:209
#20 0x00007f91df0fd259 in start_thread () from /usr/lib/libpthread.so.0
#21 0x00007f91df0095e3 in clone () from /usr/lib/libc.so.6

Thread 1 (Thread 0x7f91def08740 (LWP 164145) "mono-sgen"):
#0  0x00007f91df1098ca in __futex_abstimed_wait_common64 () from /usr/lib/libpthread.so.0
#1  0x00007f91df103270 in pthread_cond_wait@@GLIBC_2.3.2 () from /usr/lib/libpthread.so.0
#2  0x000055a80fa6088e in mono_os_cond_wait (cond=0x55a80fc2c920 <my_debug_bridge_cond>, mutex=0x55a80fc2c960 <my_debug_lock>) at ../../mono/utils/mono-os-mutex.h:219
#3  processing_stw_step () at sgen-tarjan-bridge.c:1106
#4  processing_stw_step () at sgen-tarjan-bridge.c:1039
#5  0x000055a80fa59924 in sgen_bridge_processing_stw_step () at sgen-bridge.c:227
#6  0x000055a80fa6d629 in sgen_client_bridge_processing_stw_step () at ../../mono/metadata/sgen-client-mono.h:250
#7  finish_gray_stack (generation=generation@entry=0, ctx=...) at sgen-gc.c:1191
#8  0x000055a80fa6f5c7 in collect_nursery (reason=reason@entry=0x55a80fb8dd7c "Nursery full", is_overflow=is_overflow@entry=0) at sgen-gc.c:1953
#9  0x000055a80fa734d8 in sgen_perform_collection_inner (requested_size=requested_size@entry=4096, generation_to_collect=generation_to_collect@entry=0, reason=0x55a80fb8dd7c "Nursery full", forced_serial=0, stw=stw@entry=1) at sgen-gc.c:2685
#10 0x000055a80fa73aa7 in sgen_perform_collection (stw=1, forced_serial=<optimized out>, reason=<optimized out>, generation_to_collect=0, requested_size=4096) at sgen-gc.c:2785
#11 sgen_ensure_free_space (size=4096, generation=generation@entry=0) at sgen-gc.c:2651
#12 0x000055a80fa62fd0 in sgen_alloc_obj_nolock (vtable=0x55a810e66eb8, size=112) at sgen-alloc.c:279
#13 0x000055a80fa4259b in mono_gc_alloc_vector (vtable=0x55a810e66eb8, size=112, max_length=10) at sgen-mono.c:1368
#14 0x0000000040284f0e in ?? ()
#15 0x00007f91de40a710 in ?? ()
#16 0x000055a810cc3060 in ?? ()
#17 0x0000000000000070 in ?? ()
#18 0x000055a810e66eb8 in ?? ()
#19 0x000055a810cc29f0 in ?? ()
#20 0x000055a810d0e840 in ?? ()
#21 0x00007ffc90c85670 in ?? ()
#22 0x00007ffc90c855c0 in ?? ()
#23 0x000055a800000000 in ?? ()
#24 0x000000004027c63c in ?? ()
#25 0x00007f91de40b398 in ?? ()
#26 0x000000000000000a in ?? ()
#27 0x00007f91de40a408 in ?? ()
#28 0x00007f91de40a780 in ?? ()
#29 0x00007ffc90c856b0 in ?? ()
#30 0x00007ffc90c85790 in ?? ()
#31 0x00007f91de40a500 in ?? ()
#32 0x000055a810d441f0 in ?? ()
#33 0x00007f91de40a500 in ?? ()
#34 0x00007f91de40a500 in ?? ()
#35 0x0000000000000a12 in ?? ()
#36 0x00007f91de40a6f8 in ?? ()
#37 0x00007ffc90c85790 in ?? ()
#38 0x000000004029d64c in ?? ()
#39 0x0000000000000a12 in ?? ()
#40 0x00007f91de40a640 in ?? ()
#41 0x00007f91de40a688 in ?? ()
#42 0x000000004029d564 in ?? ()
#43 0x00000000000064b4 in ?? ()
#44 0x000055a810d0e840 in ?? ()
#45 0x00007f91de40a6f8 in ?? ()
#46 0x00007f91de40a670 in ?? ()
#47 0x00007f91de40a670 in ?? ()
#48 0x000000004027b338 in ?? ()
#49 0x000055a810ccc118 in ?? ()
#50 0x00007ffc90c857f8 in ?? ()
#51 0x00007ffc90c859a0 in ?? ()
#52 0x000000004027b290 in ?? ()
#53 0x0000000000000000 in ?? ()
[Inferior 1 (process 164145) detached]

=================================================================
        Basic Fault Address Reporting
=================================================================
Memory around native instruction pointer (0x7f91def47d22):0x7f91def47d12  d2 4c 89 ce bf 02 00 00 00 b8 0e 00 00 00 0f 05  .L..............
0x7f91def47d22  48 8b 84 24 08 01 00 00 64 48 2b 04 25 28 00 00  H..$....dH+.%(..
0x7f91def47d32  00 75 1f 44 89 c0 48 81 c4 18 01 00 00 c3 48 8b  .u.D..H.......H.
0x7f91def47d42  15 09 31 18 00 f7 d8 41 b8 ff ff ff ff 64 89 02  ..1....A.....d..
Aborted (core dumped)
```

</details>

**Comment by lambdageek** [comment]
> i added a little bit of code in the sgen (`job_major_mod_union_preclean`) and tarjan bridge code (`processing_stw_step`) to add condition for them to wait for each other here (and a dummy bridge implementation):
> 
> [tmijieux@e9cd7fb](https://github.com/tmijieux/mono/commit/e9cd7fbe24502e4c57b3d6aac29cb100dc6a9b0d)
> 
> simple program.cs to reproduce bug (does meaningless stuff just to trigger gc)
> If you run this patch on this very simple csharp code, then the bug will reproduce systematically ! (on regular linux box which is more friendly to debug than android simulator/device):

@BrzVlad @steveisok is there some way that we can add a GC Bridge for testing purposes (in dotnet/runtime) so that we can have regression tests for issues like this?

**Comment by BrzVlad** [comment]
> is there some way that we can add a GC Bridge for testing purposes (in dotnet/runtime) so that we can have regression tests for issues like this?

@lambdageek @steveisok In mono/mono we had `test-sgen` and `test-sgen-bridge*` targets in `mono/tests/Makefile.am`. I know we had some plans to port our own runtime tests to dotnet/runtime. Not sure if that is still in the plans.
